hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/CloseableProcessorSupplier.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.core;
+
+import com.hazelcast.jet.function.DistributedIntFunction;
+import com.hazelcast.jet.function.DistributedSupplier;
+import com.hazelcast.logging.ILogger;
+
+import javax.annotation.Nonnull;
+import java.io.Closeable;
+import java.util.Collection;
+import java.util.stream.IntStream;
+
+import static com.hazelcast.jet.impl.util.ExceptionUtil.sneakyThrow;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * A {@link ProcessorSupplier} which closes created processor instances
+ * when the job is complete.
+ *
+ * @param <E> Processor type
+ */
+public class CloseableProcessorSupplier<E extends Processor & Closeable> implements ProcessorSupplier {
+
+    static final long serialVersionUID = 1L;
+
+    private final DistributedIntFunction<Collection<E>> supplier;
+
+    private transient ILogger logger;
+    private transient Collection<E> processors;
+
+    /**
+     * @param simpleSupplier Supplier to create processor instances.
+     *                       Parameter is the local processor index.
+     */
+    public CloseableProcessorSupplier(DistributedSupplier<E> simpleSupplier) {
+        this(count -> IntStream.range(0, count)
+                               .mapToObj(i -> simpleSupplier.get())
+                               .collect(toList()));
+    }
+
+    /**
+     * @param supplier Supplier to create processor instances.
+     */
+    public CloseableProcessorSupplier(DistributedIntFunction<Collection<E>>  supplier) {
+        this.supplier = supplier;
+    }
+
+    @Override
+    public void init(@Nonnull Context context) {
+        logger = context.logger();
+    }
+
+    @Nonnull @Override
+    public Collection<E> get(int count) {
+        assert processors == null;
+        return processors = supplier.apply(count);
+    }
+
+    @Override
+    public void complete(Throwable error) {
+        if (processors == null) {
+            return;
+        }
+
+        Throwable firstError = null;
+        // close all processors, ignoring their failures and throwing the first failure (if any)
```

Comment:
This is a self-contradictory comment. If we throw, we don't ignore. I prefer to just delete the comment, the code speaks for itself.